### DISCUSSION TITLE
REQUEST: Satellite planning

### DISCUSSION THREAD

**Discussion by ee5cd7d5d** [mworion/MountWizzard4#153] [Closed Feature requests]
Hi @mworion !
Great work with the MW4! We have been using it at the TU München for some time now and so far it's working great with our equipment.
We were discussing the idea of using it to program several satellite observations instead of tracking a single satellite each time. Something like an observation session, such that when one observation has been made then the next starts and so on. Before I start playing with the code myself I was wondering if something like that is on your roadmap? Is it a feature that you would consider in the MW4?
Thanks! And keep up the awesome work! :+1:

**Comment by mworion** [comment]
Pedro,
guess you are using a 10micron mount for satellite tracking. Otherwise It would not make sense. I would like to learn what you are doing with MW4. Right now I have to finish up the first release of 2.0. 
I though of a small planning tool, which selects a list of satellites to be tracked during the next night depending on a set of constraints and parameters. Still this was not intended to be a scheduler. But anyhow MW4 is open source and you are invited to play around with it. If you don't mind and have some parts done, you also might contribute and add some features as well.
Michel

**Reply by ee5cd7d5d** [reply]
Hi Michel
yes no rush! Thanks first of all for your quick answer. Maybe I can start by quickly explaining what we do with the MW4. I am currently leading a small group of students at the Technical University, we are using an old observatory (from the 70s) located in the center of Munich to make satellite observations, since the atmosphere makes it difficult to do other kinds of science. Our long term goals are to automate as much of the observatory as possible to be able to operate it remotely and do satellite observations in cooperation with other sites.
So actually the planning tool that you have in mind would also be very useful for us because often we spend a long time trying to filter the satellites we want to observe based on type, brightness, pass time and elevation. We usually do this manually although I was already planning to automate it. I could use that filtering tool you're planning as a starting point to create the scheduler, if that's okay with you. If a scheduler is not on your roadmap then I might develop some features myself :) although I have to carefully plan it since I don't know the MW4 in depth of course.
Any thoughts?
Pedro

**Reply by mworion** [reply]
Pedro,
as the v2 is getting close. Would you mind (if there is some time left) using the v2 beta and give me some feedback about the tracking quality of the algorithms ?
There are two in place: the mount built in where you only have to download the TLE file and secondly which is new to v2 if you are running firmware 3.0 or later: calculation of the sat track externally and transferring it to the mount. In this case, horizon, flipping etc. could be taken into account. Rüdiger, who is testing Sat tracking as well discovered some glitches in time base (GPS vs. PC controlled) in order to get a sat right on spot.
Michel

PS:
As you are in Munich, there also might be the chance to visit your site and talk.

**Reply by ee5cd7d5d** [reply]
Hi @mworion !

That would be awesome! I might be able to test this weekend... As you might now the weather in Munich is pretty bad this week, I was planning to go one of these days to do some maintenance but i won't be able to do any observations due to the weather. 

I think it's really cool that you're interested in meeting in person. Our observatory is located at the main TU Munich building in the city center, maybe we can meet there this weekend or one of the coming weeks? What do you think?

**Reply by mworion** [reply]
Coming weekend (Fri, Sat or Sun) would work for me, after that I'm on holiday. And yes it's raining all day long....

**Reply by ee5cd7d5d** [reply]
Sounds good! We're planning on going on Sunday evening, I'll post here depending on weather and stuff. Will that work for you? The TU München building is Arcisstraße 21 just to clarify :)

**Reply by ee5cd7d5d** [reply]
@mworion it seems like we will not go tomorrow due to weather conditions. You mentioned you're on vacation the coming days, should i just text here when we go one day?

**Reply by mworion** [reply]
Yes please. There are for sure many more opportunities, there is no hurry.

**Comment by mworion** [comment]
Let’s give it a try.

**Comment by mworion** [comment]
Pedro, 
in between, Could you make a list of criteria auf how you would check them against which value for the search?
I definitiv will implement some extended search and export. The scheduler itself we have to discuss. So any ideas how you would like to select would be good to know.
Michel

**Reply by ee5cd7d5d** [reply]
Hi @mworion !

First of all thanks for showing interest in our use of the application! 😊

So on a typical night of observations i would search for an overpass satellite list on a website like heavens-above.com or in-the-sky.org and then filter from that list based on three criteria: max elevation, brightness and "interestingness" 😂 by which i mean that something like an upper stage is not interesting and the ISS is the most interesting, with Earth Observation satellites ranging in between.

That means that the simplest filter criteria would be by name or catalogue number, but over the entire database. The current implementation (I have 2.0.0b8 though, i was not able to update to b44 because i made some modifications on my local install and i broke something, i will try to fix it tomorrow) pre-filters by satellite type, which is nice but usually if I want to observe a large number of satellites with passes starting every few minutes i don't have time to click through all the list of satellite types. I can talk for myself of course, other users will have other needs 😁

Ideal filter criteria would be also elevation and brightness for a given night, but calculating those for every satellite on a given night is resource intensive and would require some effort. Not to mention calculating the passes themselves, which is needed to get the list in the first place. I think the easiest implementation would be really a search as suggested above, leaving the calculation of the table of passes to another website (though it would be nice to have it all in the MW haha that goes against the Linux philosophy too -one program does one thing well-). Then a planner would simply compile a list of satellites to follow, although one should think of the logic to follow if passes overlap...

But i digress 😊 i hope that answered your question, I'll be happy to further discuss...

**Reply by mworion** [reply]
Hi, checked some filter performance in python. On my MacBook Pro I used the active payload file (7600 entries) from Sat Track. Filtering all sats which pass with at least 89 deg altitude within the next 12 hours get their rang and rage rate and checking if they are sunlit (so you could watch them) takes about 20 seconds. So it's not feasible doing this while typing, but reasonable fast. If you go below 89 degrees (89 gives 120 satellites), you get much more events, which also increase computer time. I anyway would encourage you to use the skyfield library from Brandon Rhodes (as I do in MW4). Doing the explained task is 30 lines of code. So there is a good chance to sort most things out, before really start to code.

**Reply by ee5cd7d5d** [reply]
Hi @mworion 

That sounds pretty good actually, as i mentioned before filtering by elevation and brightness or other criteria isn't even the main criterion, i think the ability to search the entire database is more relevant for my use case. For instance, i have been trying to get a hold on old satellites which aren't active anymore, since space junk is an important research topic at the moment. But maybe tomorrow i want to track a different satellite which has an active payload, like a Starlink satellite which are also interesting for being new and such a huge constellation. If i remember correctly, the MW4 stores the satellite data in files, but how would you feel about a direct query to CelesTrak? Would that be an option?

Cheers, Pedro

**Reply by mworion** [reply]
Hi Pedro,
just an update: If you already had MW4 v2.x installed, please manually select v2.0.1b1 as version and update. You might have a look to the search page on satellites. This would be my proposal on MW4 side.
You take about other databases you would like to search. If you have some more locations, where to download, please let me know. Actually I only read TLE files, but probably OMM to come.
Michel

**Reply by ee5cd7d5d** [reply]
Hi Michel,
that looks pretty good! And the performance you managed is astounding. There is always a "but" though :wink: but: when doing observations for a particular night one can only search through the satellites that have been downloaded, and there is no chance to search through all the files, which means that I need to set the file I am searching first (for example 100 Brightest satellites), and maybe the satellite I am looking for is not there, and I have to click and filter all the files. There is also no way I can get simply "all visible satellites passing within the next hours" as I can get [here](https://www.heavens-above.com/AllSats.aspx?lat=51.5&lng=0&loc=M%c3%bcnchen&alt=0&tz=CET). Some questions:

- Wouldn't it make more sense to load _everything from every file_ and then add a filter? I suppose that would heavily impact performance...
- Let's assume for instance I want to track "any Cosmos upper stage". How would I go about it?
- Alternatively, can I add custom TLE files? Like if I get a custom file from CelesTrak, can I just add it? Sorry I did not have time to look at the code in this regard :smile: 

In any case it's definitely much much better for my purposes! I will test as soon as I get a clear sky and let you know.

By database search I meant directly querying CelesTrak (see [here](https://celestrak.com/NORAD/documentation/gp-data-formats.php) ) instead of using files. What do you think of that idea?

**Reply by mworion** [reply]
Hi Pedro,
a custom file is actually not included, but simple to realize. If you have an overall DB with all sat information or additional ones to these I choose from celestrak, please let my know.
I checked against the single sat data API from celestrak, but this is hard to download many data. To just get a single sat, it might be usable, but the all the search has to be done before !

**Reply by ee5cd7d5d** [reply]
Hi Michael,
I was afraid you would say the CelesTrak API does not perform :disappointed: I unfortunately do not know of any other database. The websites I use for observations are not really databases per se, they are rather user-oriented.
Alternatively reading from a custom file would be very helpful, because then I can simply compose a file with "interesting satellites" and keep it up to date, and that would solve most of the use cases I can think of... What do you think?

**Reply by mworion** [reply]
Hi Perdo,
how many custom entries do you need ? Adding them to the list is quit easy. So I just added one "Custom", which expects a files custom.txt with the TLE entries like the others in /data directory. You might test it next beta. 
Michel

**Reply by ee5cd7d5d** [reply]
Hi Michel,
I would like to test a bit with the new custom file, but I expect if I want to do observations for a night I would need max. 100 entries. If I decide to observe for instance Starlink satellites, then one would have 1000+ satellites in the file (I think how they have like 1700...)
I think that should be alright with the performance.

**Comment by mworion** [comment]
If you right now use the active sat database, you could judge about the performance: in this file there are about 4500 satellites with its data.

**Reply by mworion** [reply]
v2.0.1b2 should do the custom file

**Reply by ee5cd7d5d** [reply]
I just tested, that should work for my purposes. Now I just need to do some web scraping and I'm good to go. Just to be sure, I just need to generate a custom.txt file and put it under "data", correct? Are you planning to change that feature? The current behavior is fine for me, I just want to know your thoughts :smile:

**Reply by mworion** [reply]
I store all downloaded TLE files under data, so I would keep the solution as well for the custom.txt file.

**Reply by ee5cd7d5d** [reply]
![IMG_20210924_224825.jpg](https://user-images.githubusercontent.com/49471546/134739561-87c83d94-6f90-4b63-923c-a2262f0fd1c7.jpg)

@mworion Works like a charm! That's one of tonight's satellites as downloaded from the custom TLEs file

**Reply by mworion** [reply]
Pedro,
added some more filters (apparent magnitude of satellite) and improved the GUI. So v2.1.0b7 will show this. Coming close to release.
Could you describe your experience in centering the sats? 
I would be happy if you could use the internal maths as well (probably for the same sat) and help me tuning this feature as well.
Michel

**Reply by ee5cd7d5d** [reply]
> Pedro,
> added some more filters (apparent magnitude of satellite) and improved the GUI. So v2.1.0b7 will show this. Coming close to release.
> Could you describe your experience in centering the sats? 
> I would be happy if you could use the internal maths as well (probably for the same sat) and help me tuning this feature as well.
> Michel

Yes I can check the new version tomorrow. I can also test the tracking if the sky is open enough. 

Could you elaborate what you mean with centering the sats? And regarding the usage of the internal maths? I'm not sure what you mean 😃 Usually i just select the target and the mount takes care of the rest. The centering of the sat is done either manually or by ensuring that the mount model is accurate enough. I thought about using a tracker, but that's far off for now, as our current priority is the automation of the dome.

**Reply by mworion** [reply]
Pedro,

During tracking you have on the right side three sliders / button setups to change tracking position in relation to satellite. On for time (meaning along the tracking path - the most important part) and two for the axis RA / DEC. This should work whenever you track a satellite. Beside the Handpad this is the way centering manually the satellite from MW4 side.

Automation of the dome should be working as well (also subject to test as I don't have a dome).

The interesting part is the internally calculation: This is done not in the mount, but MW4 calulates the satellite track of its own and transfers the data in steps to the mount, which finally follow the way points. I made a comparison between mount calculated path and MW4 calculated path. The difference is only mac 10 arc sec. There is a important "but": If I use the same time base that is the case, but when doing a model, the mount could correct it's internal time according to the real stars. This correction is not accessible from outside (e.g. MW4). This delta is also not reflected in the time gathered from the mount itself, because I guess that the corrections is applied after the mount calculated the "nominal" track and then doing the model corrections afterwards. 

This means, that MW4 tracking path is shifted along the track by +- 1s. This could be corrected with the first slider. If possible you could test this and report back if that works. If so, the assumptions were correct a I could work with them.

Why doing the MW4 calculations at all you might ask: A could add constraints in term of avoid flips, repsect horizon lines etc. in the calculation whereas the mount itself does not.
Michel

**Comment by mworion** [comment]
Planning is now out in v2.1x

**Reply by ee5cd7d5d** [reply]
Fantastic! Will test it today. Weather should be good. By the way, it seems like it's a little difficult to allow someone external to the TU Munich in the observatory :( but I will try to organize a visit concept. It's also more difficult because of corona... :(

**Reply by ee5cd7d5d** [reply]
![IMG_20211203_203447.jpg](https://user-images.githubusercontent.com/49471546/144662427-00fb01ac-4b8d-4c25-9021-80da260739df.jpg)

Hmmm i can't find any planning tab? Or is it integrated in the current ones and I'm just being dumb? 🤣

**Reply by mworion** [reply]
Hi Pedron,
sorry to tell I had to revert the sat scheduler as it broke in some areas the UI event queue. The idea was to select a sat from the list and add it to the scheduler. The scheduler sorts the sat passes by time and you could start it / stop it. But that's it.
Any comments ?

For a possible visit: Please take your time. I'm definitely interested in visiting the site, but it should not be stressful :-)
Michel
